### PR TITLE
Sprint 7 Release

### PR DESCRIPTION
**Pull request by lpendem**
# Sprint 7 Release including:
## [v0.0.14](https://github.com/nci-ats/fs-middlelayer-api/releases/tag/v0.0.14)
#207 Create command to generate user and admin accounts
#223 Fix swagger validation issues
#301 ContId required in contact/person and org post route
#307 Rename derefSchema to validationSchema
#314 Change search contact Basic API
#318 Remove the "contact" field from the /contact/person POST route
#319 Remove trailing slash on /contact-address basic api route
#320 Remove district from securityID
#321 Convert area codes and phone numbers to strings when posting to basic url
#322 Remove trailing slash from /application POST
#323 Change ePermitID capitalization
#325 formName for noncommercial to "FS-2700-3B"
#324 Autogenerate epermitIDs
#326 add purpose to the noncommercial fields post to application
#329 Update start and end time validation
#333 conditional contact/contCn carry over in postRequest
#334 Remove the random control numbers from the index.js gets
#335 Handle contact found
#336 If the contact cn is found the promise
#337 Post to org when organization Type is not individual

## [v0.0.15](https://github.com/nci-ats/fs-middlelayer-api/releases/tag/v0.0.15)
#254 Research and implement object clone
#296 Audit Swagger for accessibility
#297 Handle 404 from Basic API
#328 Update Swagger display data model
#330 Update swagger to show body request
#352 Post an organization to /contact/organization
#353 Upgrade deref from 0.6.4 to 0.7.0

### CODE REVIEW

File: src/controllers/basic.js [hunk 1]
```diff
@@ -37,98 +85,148 @@ function getAutoPopulatedFields(basicFields){
...
+ * @param  {Array} fieldsToBasic - All fields in object form which will be sent to basicAPI
  * @param  {Object} body - user input
  * @param  {Object} autoPopValues - All values which have been auto-populated
- * @return {Array} - Array of post objects
+ * @return {Object} - Array of post objects
  */
-function getBasicFields(fields, body, autoPopValues){
-	const requests = [], postObjs = [];
-	fields.forEach((field)=>{
+function getBasicFields(fieldsToBasic, body, autoPopValues){
```
Comment (lauraGgit) [inline src/controllers/basic.js]: this function seems like it could be broken down to reduce complexity.

File: src/controllers/createUser.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
+/*
+
+  ___ ___       ___               _ _       _   ___ ___ 
+ | __/ __|  ___| _ \___ _ _ _ __ (_) |_    /_\ | _ \_ _|
+ | _|\__ \ / -_)  _/ -_) '_| '  \| |  _|  / _ \|  _/| | 
+ |_| |___/ \___|_| \___|_| |_|_|_|_|\__| /_/ \_\_| |___|
+
+*/
+
+//*******************************************************************
```
Comment (lauraGgit) [inline src/controllers/createUser.js]: This looks easy to maintain and super useful.

File: src/controllers/createUser.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
+/*
+
+  ___ ___       ___               _ _       _   ___ ___ 
+ | __/ __|  ___| _ \___ _ _ _ __ (_) |_    /_\ | _ \_ _|
+ | _|\__ \ / -_)  _/ -_) '_| '  \| |  _|  / _ \|  _/| | 
+ |_| |___/ \___|_| \___|_| |_|_|_|_|\__| /_/ \_\_| |___|
+
+*/
+
```
Comment (lauraGgit) [inline src/controllers/createUser.js]: This looks easy to maintain and super useful.

File: src/controllers/basic.js [hunk 4]
```diff
@@ -18,12 +18,60 @@ const request = require('request-promise');
 //*******************************************************************
 // other files
 
+const error = require('./error.js');
 const db = require('./db.js');
+const DuplicateContactsError = require('./duplicateContactsError.js');
 const SUDS_API_URL = process.env.SUDS_API_URL;
 
 //*******************************************************************
+// AUTO-POPULATE FUNCTIONS
```
Comment (lauraGgit) [inline src/controllers/basic.js]: I'd rather have the auto-populate functions be another file.

File: src/controllers/basic.js [hunk 5]
```diff
@@ -227,43 +327,92 @@ function createApplication(fieldsObj, contCN, postObject){
...
+}
+
 /** Sends requests needed to create an application via the Basic API
  * @param  {Object} req - Request Object
  * @param  {Object} res - Response Object
- * @param  {Object} sch - Schema object 
+ * @param  {Object} sch - Schema object
  * @param  {Object} body - User input
  */
 function postToBasic(req, res, sch, body){ //Should remove control number once we get from BASIC api
```
Comment (lauraGgit) [inline src/controllers/basic.js]: This seems to have a pretty high complexity/ and seems like you could pull out some of these conditionals into its own function.

File: src/controllers/index.js [hunk 6]
```diff
@@ -233,7 +221,7 @@ const getControlNumber = function(req, res, reqData){
 	else {
```
Comment (lauraGgit) [inline src/controllers/index.js]: Seems like the two major handlers of the conditional could be broken out into separate handler functions.

File: test/test.js [hunk 7]
```diff
@@ -16,12 +16,18 @@
...
 const util = require('./utility.js');
 
 const chai = require('chai');
 const expect = chai.expect;
-const factory = require('unionized');
+const bcrypt = require('bcrypt-nodejs');
+const db = include('src/controllers/db.js');
+const models = include('src/models');
+
+const adminUsername = 'admin' + (Math.floor((Math.random() * 1000000) + 1)).toString();
```
Comment (lauraGgit) [inline test/test.js]: Seems like this admin/user generator could be moved into a helper class and then initialized on each test file.

File: src/controllers/basic.js [hunk 8]
```diff
@@ -18,12 +18,60 @@ const request = require('request-promise');
...
+// AUTO-POPULATE FUNCTIONS
+/**
+ * Concats all indexs of input
+ * @param  {Array} input - Array of strings to be joined together
+ * @return {String}      - Single string made up of all indicies of input 
+ */
+function concat(input){
+	const output = input.join('');
+	return output;
+}
```
Comment (cameronwolf) [inline src/controllers/basic.js]: Yes, being used for purpose, securityId, managingID, and adminOrg.

Comment (lauraGgit) [inline src/controllers/basic.js]: Sorry, I meant to say the noncommericial purpose generation?

**Pull request merged**